flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,533 @@
+    *
+    * @param timestamp the timestamp of the timer
+    * @param ctx       the context to register timer or get current time
+    * @param out       the collector for returning result values
+    */
+  override def onTimer(
+    timestamp: Long,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+    out: Collector[CRow]): Unit = {
+    getCurrentOperatorTime(ctx)
+    //TODO In the future, we should separate the left and right watermarks. Otherwise, the
+    //TODO registered timer of the faster stream will be delayed, even if the watermarks have
+    //TODO already been emitted by the source.
+    if (leftTimerState.value == timestamp) {
+      val rightExpirationTime = leftOperatorTime - rightRelativeSize - allowedLateness - 1
+      removeExpiredRecords(
+        timestamp,
+        rightExpirationTime,
+        leftOperatorTime,
+        rightCache,
+        leftTimerState,
+        ctx,
+        false
+      )
+    }
+
+    if (rightTimerState.value == timestamp) {
+      val leftExpirationTime = rightOperatorTime - leftRelativeSize - allowedLateness - 1
+      removeExpiredRecords(
+        timestamp,
+        leftExpirationTime,
+        rightOperatorTime,
+        leftCache,
+        rightTimerState,
+        ctx,
+        true
+      )
+    }
+  }
+
+  /**
+    * Remove the expired records. Register a new timer if the cache still holds records
+    * after the cleaning up.
+    */
+  private def removeExpiredRecords(
+    timerFiringTime: Long,
+    expirationTime: Long,
+    oppositeWatermark: Long,
+    recordCache: MapState[Long, JList[Row]],
+    timerState: ValueState[Long],
+    ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+    removeLeft: Boolean): Unit = {
+
+    val keysIterator = recordCache.keys().iterator()
+
+    // Search for expired timestamps.
+    // If we find a non-expired timestamp, remember the timestamp and leave the loop.
+    // This way we find all expired timestamps if they are sorted without doing a full pass.
+    var earliestTimestamp: Long = -1L
+    var recordTime: Long = 0L
+    while (keysIterator.hasNext) {
+      //TODO The "short-circuit" code was commented, because when using a StateMap with
+      //TODO unordered keys, the cache will grow indefinitely!
+      // && earliestTimestamp < 0) {
+      recordTime = keysIterator.next
+      if (recordTime <= expirationTime) {
+        // TODO Not sure if we can remove records directly.
+        keysIterator.remove()
+      } else {
+        // We find the earliest timestamp that is still valid.
+        if (recordTime < earliestTimestamp || earliestTimestamp < 0) {
+          earliestTimestamp = recordTime
+        }
+      }
+    }
+    // If the cache contains non-expired timestamps, register a new timer.
+    // Otherwise clear the states.
+    if (earliestTimestamp > 0) {
+      registerCleanUpTimer(ctx, earliestTimestamp, oppositeWatermark, timerState, removeLeft, false)
```

Comment:
clean up is quite expensive because we iterate over the full map. By scheduling a timer to the smallest not expired timestamp, we will do this soon again. This might be a problem in case of large window sizes. Does it make sense to delay the clean up by adding a half window size or setting the timer to the average of min and max timestamp?

Reply:
That's indeed a critical problem. The current approach is designed for minimizing the cached records, but may spend too much time in cleaning up. I think of two ways to improve that: (1) as you suggested, we delay the timers with specific spans, or (2) we schedule the cleaning up process with a specific interval according to the relative window size. What do you think?
Actually, every problem will be solved if we got a state with ordered keys, like the TreeMap ;-)

Reply:
Right, I think for now, I'd add a window size to the `earliestTimestamp` for the clean up.
In addition, we can add a check to the cache iteration in `processElement()` that compares the cached records against a cleanup threshold (which is based on the current watermark, window size, and allowedLateness). Then we can also cleanup during regular processing and the cleanup timer would be a safety net.

Since we need to go over the full cache for joining anyway, we can also cleanup at that point.